### DISCUSSION TITLE
Select file uniquely based on contents

### DISCUSSION THREAD

**Discussion by shivangp76** [junegunn/fzf#3898] [Q&A] [answered]
Hi,

I am trying to make a notes selector. I have a directory where each file is a note. I want to be able to select multiple ones that match a search query (fuzzy) and output them. The catch is that I want the selecting to be unique for the file. For example, suppose I search query A and then select a file from the results. Then, I search query B which matches the same file at a different line number. This file should already be selected at this point since I selected it with query A.

Here is my attempt so far:
```bash
  INITIAL_QUERY="${*:-}"
  fzf --multi --ansi --query "$INITIAL_QUERY" \
      --bind "start:reload:$RG_PREFIX {q}" \
      --color "hl:-1:underline,hl+:-1:underline:reverse" \
      --prompt 'notes-selector:fzf> ' \
      --delimiter : \
      --preview 'bat --color=always {1} --highlight-line {2}' \
      --preview-window 'up,60%,border-bottom,+{2}+3/3,~3' |
  awk -F: '!seen[$1]++ {printf "%s ", $1}' | sed 's/ $//'
```
The `awk` command makes the output unique, but that is not easily visible in the fuzzy selector.  I tried binding the changes to ripgrep, like in [ADVANCED.md](https://github.com/junegunn/fzf/blob/master/ADVANCED.md#switching-between-ripgrep-mode-and-fzf-mode-using-a-single-key-binding), however, then `--multi` does not work since all of the options are reset on change. This is also not fuzzy. I also tried using the flag `--max-count=1` for ripgrep, but then only the first line shows since that is the first line matching the initial query of an empty string.

I guess the ideal workflow would be that fzf only shows me the filename (not the line containing the match), but the preview window shows the line containing the match. Then it should be much easier to get `--multi` to persist across searches since the exact line will be the same (just the filename).

Any help would be appreciated. Thank you in advance!

**Comment by LangLangBart** [comment, marked as answer]
> For example, suppose I search query A and then select a file from the results. Then, I search query B which matches the same file at a different line number. This file should already be selected at this point since I selected it with query A.


An action as you described, to select all entries in the list where the first field is the same as in the currently selected line, is not an action currently present.


Alternatively, you could try to modify your function, and on every <kbd>⇥ Tab</kbd> press, the preview is altered to show your current selection. Upon triggering the `focus` event (a vertical cursor movement or a search result update), the preview switches back. When you hit <kbd>⏎ Enter</kbd>, a unique list of selected files will be printed out.


The weird `{+f1}` placeholder is described in the tables below:

|     Placeholder      |            Description            |
| :------------------: | :-------------------------------: |
|         `{}`         |           current line            |
| `{2}` |  current line with positional field 2  |
| `{2..}` |  current line from the 2nd field to the last field |
|        `{n}`         | current line number (zero index)  |
| `{q}`  |           current query           |


| Placeholder expression flags |                Description                 |
| :--------------------------: | :----------------------------------------: |
|             `+`              | space-separated list of the selected lines, when used with `f` newline separated |
|             `f`              |               temporary file               |
|             `s`              |            preserve whitespace             |



```sh
fzf --multi --ansi --query "${*:-}" \
  --bind "start:reload:rg --line-number --no-heading --color always {q}" \
  --color "hl:-1:underline,hl+:-1:underline:reverse" \
  --prompt 'notes-selector:fzf> ' \
  --delimiter : \
  --preview 'bat --color=always {1} --highlight-line {2}' \
  --preview-window 'up,60%,border-bottom,+{2}+3/3,~3' \
  --bind 'enter:become:sort -u {+f1}' \
  --bind 'tab:toggle+change-preview-window(~0:+1)+change-preview:echo "SELECTED FILES:"; sort -u {+f1} | sed "s/^/  - /"' \
  --bind 'focus:+change-preview-window(+{2}+3/3)+change-preview:bat --color=always {1} --highlight-line {2}'
```

<img src="https://github.com/junegunn/fzf/assets/92653266/4cbde6bc-6c79-4f86-993a-7b3fa6e77770" width="600">

PS: The function doesn't fulfil all your requirements, but it does provide a clearer picture of what was already selected when making a new selection. Coupled with a functioning short-term memory, it might be good enough.

**Reply by shivangp76** [reply]
Interesting idea. Thank you for the help! Using your solution, I can also press Tab twice to display the selected files and manually inspect to see if a file is selected.
